src/mscorlib/shared/System/String.cs

```diff
@@ -2,109 +2,391 @@
-    using Microsoft.Win32;
-    using System.Diagnostics;
-    using System.Security;
-
-    //
-    // For Information on these methods, please see COMString.cpp
-    //
     // The String class represents a static string of characters.  Many of
     // the String methods perform some type of transformation on the current
-    // instance and return the result as a new String. All comparison methods are
-    // implemented as a part of String.  As with arrays, character positions
-    // (indices) are zero-based.
+    // instance and return the result as a new String.  As with arrays, character
+    // positions (indices) are zero-based.
 
     [Serializable]
     [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
-    public sealed partial class String : IComparable, ICloneable, IConvertible, IEnumerable
-        , IComparable<String>, IEnumerable<char>, IEquatable<String>
+    public sealed partial class String : IComparable, IEnumerable, IEnumerable<char>, IComparable<String>, IEquatable<String>, IConvertible, ICloneable
     {
-        //
-        //NOTE NOTE NOTE NOTE
-        //These fields map directly onto the fields in an EE StringObject.  See object.h for the layout.
-        //
-        [NonSerialized] private int m_stringLength;
+        // String constructors
+        // These are special. The implementation methods for these have a different signature from the
+        // declared constructors.
 
-        // For empty strings, this will be '\0' since
-        // strings are both null-terminated and length prefixed
-        [NonSerialized] private char _firstChar;
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        public extern String(char[] value);
 
-        // The Empty constant holds the empty string value. It is initialized by the EE during startup.
-        // It is treated as intrinsic by the JIT as so the static constructor would never run.
-        // Leaving it uninitialized would confuse debuggers.
-        //
-        //We need to call the String constructor so that the compiler doesn't mark this as a literal.
-        //Marking this as a literal would mean that it doesn't show up as a field which we can access 
-        //from native.
-        public static readonly String Empty;
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private string Ctor(char[] value)
+        {
+            if (value != null && value.Length != 0)
+            {
+                string result = FastAllocateString(value.Length);
+                unsafe
+                {
+                    fixed (char* dest = &result._firstChar, source = value)
+                    {
+                        wstrcpy(dest, source, value.Length);
+                    }
+                }
+                return result;
+            }
+            else
+                return Empty;
```

Comment:
Nit: in some places we return Empty at the end versus earlier on.  It might be nice both for consistency around that and to avoid having more code indented then necessary to invert the condition, e.g.
```C#
if (value == null || value.Length == 0)
{
    return Empty;
}

string result = FastAllocateString(value.Length);
...
```
etc., here and elsewhere.  Not a big deal, though.

Reply:
I have actually changed it as you have suggested, but then changed it back because of I was not sure whether the unnatural ordering is intentional micro-optimization for "forward branch predicted not taken".

Reply:
Ok.  Maybe something to follow-up on.

Reply:
Actually, the flow gets fixed up to the right shape via IBC. So it should be fine to do this.